packages/network/src/connection/WebRtcEndpoint.ts

```diff
@@ -231,20 +227,108 @@ export class WebRtcEndpoint extends EventEmitter implements IWebRtcEndpoint {
+    }
+    
+    private onRtcOffer({ routerId, originatorInfo, description, connectionId }: OfferOptions): void {
+        const { peerId } = originatorInfo
+        
+        let connection: Connection
+       
+        if (!this.connections[peerId]) {
+            connection = this.createConnection(peerId, routerId, false)
+            
+            try {
+                connection.connect()
+            }
+            catch(e) {
+                this.logger.warn(e)
+            }
+
+            this.connections[peerId] = connection
+        }
+        else if (this.connections[peerId] && this.connections[peerId].getConnectionId() !== 'none') {
+            const conn = this.connections[peerId]
+            const lastState = conn.getLastState()
+            this.logger.trace('onRtcOffer: answering reconnect attempt; there is already a connection for %s. state: %s', 
+                NameDirectory.getName(peerId), lastState)
+            
+            connection = this.createConnection(peerId, routerId, false)
+            if (conn.getDeferredConnectionAttempt()) {
+                connection.setDeferredConnectionAttempt(conn.stealDeferredConnectionAttempt())
+            }
+            
+            delete this.connections[peerId]
+            conn.close()
+
+            try {
+                connection.connect()
+            }
+            catch(e) {
+                this.logger.warn(e)
+            }
+            this.connections[peerId] = connection
+        }
+        else {
+            connection = this.connections[peerId]
+        }
+       
+        connection.setPeerInfo(PeerInfo.fromObject(originatorInfo))
+        connection.setConnectionId(connectionId)
+
+        connection.setRemoteDescription(description, 'offer' as DescriptionType.Offer)
+    }
+
+    async connect(
+        targetPeerId: string,
+        routerId: string,
+        trackerInstructed = true
+    ): Promise<string> {
+        const offering = this.peerInfo.peerId < targetPeerId
+        const peerType = offering ? 'offerer' : 'answerer'
+
+        // Prevent new connections from being opened when WebRtcEndpoint has been closed
+        if (this.stopped) {
+            return Promise.reject(new WebRtcError('WebRtcEndpoint has been stopped'))
+        }
+
+        if (this.connections[targetPeerId]) {
+            const connection = this.connections[targetPeerId]
+            const lastState = connection.getLastState()
+            this.logger.trace('%s has already connection for %s. state: %s', peerType, NameDirectory.getName(targetPeerId), lastState)
+            if (lastState === 'connected') {
+                return Promise.resolve(targetPeerId)
+            }
+
+            if (connection.getDeferredConnectionAttempt()) {
+                return (connection.getDeferredConnectionAttempt()?.connectionAttemptPromise as Promise<string>)
+            }
+            else {
+                throw new Error(`disconnected ${connection.getPeerId()}`)
```

Comment:
In this method, the user of this class is asking for a new connection. If an existing connection existed but has disconnected, shouldn't we form a new one instead of throwing an error regarding the previous disconnection?

Reply:
I clarified the "if then else" logic here a bit. If connection is already connected, we return it. If connection is in process of connecting we return the deferred promise. Else we throw an error, because the connection should not exist in this.connections[targetPeerId] if it has already been closed (would indicate bug if it existed)